src/java/org/apache/cassandra/service/snapshot/SnapshotLoader.java

```diff
@@ -0,0 +1,141 @@
+    }
+
+    public SnapshotLoader(String[] dataDirs)
+    {
+        this.dataDirectories = Arrays.asList(dataDirs).stream().map(d -> Paths.get(d)).collect(Collectors.toList());
+    }
+
+    public SnapshotLoader(Collection<Path> dataDirs)
+    {
+        this.dataDirectories = dataDirs;
+    }
+
+    public Set<TableSnapshot> loadSnapshots()
+    {
+        if (!walked) {
+            for (Path dataDir : dataDirectories)
+            {
+                try
+                {
+                    Files.walkFileTree(dataDir, Collections.EMPTY_SET, 5, this);
+                }
+                catch (IOException e)
+                {
+                    throw new RuntimeException(String.format("Error while loading snapshots from %s", dataDir));
+                }
+            }
+            walked = true;
+        }
+        return snapshots.values().stream().map(s -> s.build()).collect(Collectors.toSet());
+    }
+
+    public FileVisitResult preVisitDirectory(Path subdir, BasicFileAttributes attrs)
+    {
+        if (subdir.getParent().getFileName().toString().equals(SNAPSHOT_SUBDIR))
+        {
+            logger.trace("Processing directory " + subdir);
+            Matcher snapshotDirMatcher = SNAPSHOT_DIR_PATTERN.matcher(subdir.toString());
+            if (snapshotDirMatcher.find())
+            {
+                try
+                {
+                    loadSnapshotFromDir(snapshotDirMatcher, subdir);
+                } catch (Throwable e)
+                {
+                    logger.warn("Could not load snapshot from %s.", subdir, e);
+                }
+            }
+            return FileVisitResult.SKIP_SUBTREE;
+        }
+
+        return subdir.equals(Directories.BACKUPS_SUBDIR)
+               ? FileVisitResult.SKIP_SUBTREE
+               : FileVisitResult.CONTINUE;
+    }
+
+    private void loadSnapshotFromDir(Matcher snapshotDirMatcher, Path snapshotDir)
+    {
+        String keyspaceName = snapshotDirMatcher.group("keyspace");
+        String tableName = snapshotDirMatcher.group("tableName");
+        UUID tableId = parseUUID(snapshotDirMatcher.group("tableId"));
+        String tag = snapshotDirMatcher.group("tag");
+        String snapshotId = buildSnapshotId(keyspaceName, tableName, tableId, tag);
+        TableSnapshot.Builder builder = snapshots.computeIfAbsent(snapshotId, k -> new TableSnapshot.Builder(keyspaceName, tableName, tableId, tag));
+        builder.addSnapshotDir(new File(snapshotDir));
+    }
+
+    /**
+     * Given an UUID string without dashes (ie. c7e513243f0711ec9bbc0242ac130002)
+     * return an UUID object (ie. c7e51324-3f07-11ec-9bbc-0242ac130002)
+     */
+    protected static UUID parseUUID(String uuidWithoutDashes) throws IllegalArgumentException
+    {
+        assert uuidWithoutDashes.length() == 32 && !uuidWithoutDashes.contains("-");
+        String dashedUUID = uuidWithoutDashes.replaceFirst("([0-9a-f]{8})([0-9a-f]{4})([0-9a-f]{4})([0-9a-f]{4})([0-9a-f]+)", "$1-$2-$3-$4-$5");
```

Comment:
you might replace this by compiled Pattern

Reply:
I could, but I prefer not to touch this right now as I'm not really sure how to use `Pattern` here and this is working as expected. Feel free to offer a suggestion on how to improve this.